frame/staking/src/lib.rs

```diff
@@ -2247,109 +2280,134 @@ impl<T: Config> Module<T> {
 	///
-	/// This will also process the election, as noted in [`process_election`].
-	fn enact_election(current_era: EraIndex) -> Option<Vec<T::AccountId>> {
-		T::ElectionProvider::elect()
-			.map_err(|e| {
-				log!(warn, "election provider failed due to {:?}", e)
+	/// In case a new era is planned, the new validator set is returned.
+	fn try_trigger_new_era(start_session_index: SessionIndex, is_genesis: bool) -> Option<Vec<T::AccountId>> {
+		let (election_result, weight) = if is_genesis {
+			T::GenesisElectionProvider::elect().map_err(|e| {
+				log!(warn, "genesis election provider failed due to {:?}", e);
+				Self::deposit_event(RawEvent::StakingElectionFailed);
 			})
-			.and_then(|(res, weight)| {
-				<frame_system::Pallet<T>>::register_extra_weight_unchecked(
-					weight,
-					frame_support::weights::DispatchClass::Mandatory,
-				);
-				Self::process_election(res, current_era)
+		} else {
+			T::ElectionProvider::elect().map_err(|e| {
+				log!(warn, "election provider failed due to {:?}", e);
+				Self::deposit_event(RawEvent::StakingElectionFailed);
 			})
-			.ok()
-	}
+		}
+		.ok()?;
 
-	/// Process the output of the election.
-	///
-	/// This ensures enough validators have been elected, converts all supports to exposures and
-	/// writes them to the associated storage.
-	///
-	/// Returns `Err(())` if less than [`MinimumValidatorCount`] validators have been elected, `Ok`
-	/// otherwise.
-	pub fn process_election(
-		flat_supports: frame_election_provider_support::Supports<T::AccountId>,
-		current_era: EraIndex,
-	) -> Result<Vec<T::AccountId>, ()> {
-		let exposures = Self::collect_exposures(flat_supports);
-		let elected_stashes = exposures.iter().cloned().map(|(x, _)| x).collect::<Vec<_>>();
+		<frame_system::Pallet<T>>::register_extra_weight_unchecked(
+			weight,
+			frame_support::weights::DispatchClass::Mandatory,
+		);
 
-		if (elected_stashes.len() as u32) < Self::minimum_validator_count().max(1) {
+		let exposures = Self::collect_exposures(election_result);
+
+		if (exposures.len() as u32) < Self::minimum_validator_count().max(1) {
 			// Session will panic if we ever return an empty validator set, thus max(1) ^^.
-			if current_era > 0 {
-				log!(
+			match CurrentEra::get() {
+				Some(current_era) if current_era > 0 => log!(
 					warn,
-					"chain does not have enough staking candidates to operate for era {:?} ({} elected, minimum is {})",
-					current_era,
-					elected_stashes.len(),
+					"chain does not have enough staking candidates to operate for era {:?} ({} \
+					elected, minimum is {})",
+					CurrentEra::get().unwrap_or(0),
+					exposures.len(),
 					Self::minimum_validator_count(),
-				);
+				),
+				None => {
+					// The initial era is allowed to have no exposures.
+					// In this case the SessionManager is expected to choose a sensible validator
+					// set.
+					// TODO: this should be simplified #8911
```

Comment:
is this a blocker?

Reply:
it is not different from before the PR, I think it is ok to fix in a follow-up